File: challenge-3/src/validate_keys.py
def is_valid_password(policy: tuple, password: int) -> int:
    min_count, max_count, char = policy
    count = password.count(char)
    return min_count <= count <= max_count


def find_invalid_key(data) -> list:
    invalid_keys = []

    for line in data:
        policy_str, password = line.split(':')
        password = password.strip()

        counts, char = policy_str.split(' ')
        min_count, max_count = map(int, counts.split('-'))

        policy = (min_count, max_count, char)

        if not is_valid_password(policy, password):
            invalid_keys.append(password)

    return invalid_keys


def main() -> None:
    with open('../data/encryption_policies.txt', 'r') as file:
        data = file.readlines()

    invalid_keys = find_invalid_key(data)

    if len(invalid_keys) >= 42:
        print(f"The 42nd invalid key is: {invalid_keys[41]}")
    else:
        print("There are fewer than 42 invali keys.")


if __name__ == "__main__":
    main()

File: challenge-3/validateKeys.ts
import { readFileSync } from "node:fs";

interface Policy {
  minCount: number;
  maxCount: number;
  char: string;
}

function isValidPassword(policy: Policy, password: string): boolean {
  const { minCount, maxCount, char } = policy;
  const count: number = password.split(char).length - 1;
  return minCount <= count && count <= maxCount;
}

function findInvalidKey(data: string[]): string {
  const invalidKey: string[] = [];

  data.forEach((line: string): void => {
    const [policyStr, password] = line.split(":");
    const trimmedPassword: string = password.trim();

    const [counts, char] = policyStr.split(" ");
    const [minCount, maxCount] = counts.split("-").map(Number);

    const policy: Policy = { minCount, maxCount, char };

    if (!isValidPassword(policy, trimmedPassword)) {
      invalidKey.push(trimmedPassword);
    }
  });

  return invalidKey[41];
}

function main() {
  const data: string[] = readFileSync(
    "./data/encryption_policies.txt",
    "utf-8"
  ).split("\n");

  const invalidKey: string = findInvalidKey(data);

  if (invalidKey) {
    console.log(`The 42nd invalid key is: ${invalidKey}`);
  } else {
    console.log("There are fewer than 42 invalid keys.");
  }
}

main();
